various improvements/corrections to the multisegment well model

GitPaean opened the pull request:
And also trying to activate the multisegment well model by default. 

With multisegment well model, more parameters can be tuned and tested, there might be some small updates base on the testing result. 

Creating this PR to warp up the efforts to improve the multisegment well model for now.

GitPaean committed:
using a small value drawdown for connection

when crossflow is banned from the well, while the crossflow is happening
for this connection for mutlisegment well.

GitPaean committed:
UseMultisegmentWell is defaulted to be true

to make the testing of the multisegment well model easier.

GitPaean committed:
considering the pressure dependence of the segment volume

segment volume is fixed, while duo to composition and pressure change,
within the fixed segment volume, different amounts of fluids can be
contained.

GitPaean committed:
handling the situation that not all the connections are open

Then Wells struct and well from parser hold different numbers of
connections.

The current approach is a temporary solution. The final approach should
be done through parser side.

Things can be difficult is because that connections can be removed due
to multiple different causes after parsing. How to keep information
always updated is not trivial.

GitPaean committed:
updating the primary variables before calculating accumulation term

GitPaean committed:
using B_avg for the inner iteration for multisegment wells

GitPaean committed:
fixing mistakes due to rebasing.

GitPaean committed:
various improvements to the inner iteration of ms wells

GitPaean committed:
changing the sign of the mass balance equations for ms wells

following the conventions with StandardWell

GitPaean committed:
setting the simulation parameters

GitPaean commented:
With multisegment well model, the relaxation of tolerance (see https://github.com/OPM/opm-simulators/pull/1584) impacts the performance in a significant way.  We are at the point to make a decision about the tolerance or we try to choose different criteria for the convergence checking. 

At the moment, the required tolerance is affected by the compressibility of the field.

GitPaean commented:
jenkins build this please

alfbr commented:
benchmark

ytelses commented:
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=286

GitPaean commented:
benchmark

ytelses commented:
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=288

alfbr commented:
Seems there was a hick-up, let's try again:
benchmark

alfbr commented:
and one more time:
benchmark

ytelses commented:
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=289

ytelses commented:
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=290

ytelses commented:
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.965    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.954    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.996    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 1.034    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=292

alfbr commented:
Please rebase, we need to build it again for testing.

GitPaean committed:
fixing some mistakes from rebasing

GitPaean commented:
done rebasing!

jenkins build this please

GitPaean commented:
jenkins build this please

GitPaean committed:
fixing the segment treatment in the init of WellStateFullyImplicitBlackoil

GitPaean committed:
first version of restarting for multisegment well

GitPaean committed:
cleaning up and fixing the compilation error from rebasing

GitPaean committed:
correcting of the sign when calculating perforation pressure

the sign of the perforation and cell pressure difference was wrong.

GitPaean committed:
first version of using upwinding to handle mass balance for MSW

some details to come to make it complete.

We are not handling it for injectors and also we are not considering the
upwinding when handle the frictional pressure loss and acceleration
pressure loss.

GitPaean committed:
adding residual of control equation to getWellResiduals() for MSW

then it is consistent that when it is converged, we do not need to enter
this function anymore.

GitPaean committed:
setting UseMultisegmentWell to be false by default

GitPaean commented:
jenkins build this update_data please

GitPaean commented on opm/autodiff/BlackoilWellModel.hpp:
```diff
@@ -358,7 +358,7 @@ namespace Opm {
 
 
-            void computeAverageFormationFactor(std::vector<double>& B_avg) const;
+            void computeAverageFormationFactor(std::vector<Scalar>& B_avg) const;
```
I do not recall which function made me change the `double` to `Scalar` here. Hopefully it is not a bid deal.

GitPaean committed:
some minor cleaning up

GitPaean commented:
jenkins build this update_data please

GitPaean commented:
jenkins build this opm-tests=109 please

GitPaean committed:
upwinding segment for top segment is itself

GitPaean commented:
jenkins build this opm-tests=109 please

GitPaean commented:
jenkins build this opm-tests=109 please

GitPaean commented:
jenkins build this update_data please

GitPaean committed:
silencing some compilation warnings.

GitPaean commented:
jenkins build this opm-tests=109 please

GitPaean commented:
I am done updating.  This is ready to go in now. 

The rest part of the development will be proposed as a new PR.

akva2 commented:
please merge so i can get the packages out the door.

GitPaean commented:
This PR has been tested with ensemble, it significantly improves the capacity of the MSW. There are some more development coming as a separate PR.  Merging it now for the release.
